src/lib/krb5/ccache/cc_file.c

```diff
@@ -765,19 +765,35 @@ fcc_next_cred(krb5_context context, krb5_ccache id, krb5_cc_cursor *cursor,
         goto cleanup;
     file_locked = TRUE;
 
-    /* Load a marshalled cred into memory. */
-    ret = get_size(context, fcursor->fp, &maxsize);
-    if (ret)
-        goto cleanup;
-    ret = load_cred(context, fcursor->fp, fcursor->version, maxsize, &buf);
-    if (ret)
-        goto cleanup;
-    ret = k5_buf_status(&buf);
-    if (ret)
-        goto cleanup;
+    while (1) {
+        /* Load a marshalled cred into memory. */
+        ret = get_size(context, fcursor->fp, &maxsize);
+        if (ret)
+            goto cleanup;
+        ret = load_cred(context, fcursor->fp, fcursor->version, maxsize, &buf);
+        if (ret)
+            goto cleanup;
+        ret = k5_buf_status(&buf);
+        if (ret)
+            goto cleanup;
+
+        /* Unmarshal it from buf into creds. */
+        ret = k5_unmarshal_cred(buf.data, buf.len, fcursor->version, creds);
+        if (ret)
+            goto cleanup;
 
-    /* Unmarshal it from buf into creds. */
-    ret = k5_unmarshal_cred(buf.data, buf.len, fcursor->version, creds);
+        /*
+         * Skip creds that have been removed by fcc_remove_cred.  Avoid skipping
+         * creds that were never valid (such as gssproxy's), which need to appear
+         * in listings.
```

Comment:
I think endtime == authtime == 0 also applies to config entries.

The reversed nature of the conditional makes this code a little hard to read, so we might want a trivial predicate function for "cred is removed".

Reply:
gssproxy stores an encrypted copy of the actual credentials in the ccache; this prevents the applications from actually using them directly.  I'm unsure if there's a reason the times being 0 is important; maybe @simo5 can help with that?

Reply:
I do not recall if the times were important, sorry.

Reply:
It's a step beyond Heimdal, but we could set authtime=~0 and endtime=0 when deleting creds.  That combination should be safe to skip, since psuedo-creds such as config entries should generally have authtime=0 and real creds should have endtime!=0.